### PR TITLE
Getteacher

### PR DESCRIPTION
**Pull request by JonathanWacks**
# Description

Updated CRM integration to use Query Expression and aliased values - this allows us to more easily pull back (most of) the data in one request.  The qualifications are retrieved separately, to reduce data duplication, as there could be multiple qualifications for the teacher.

Trello cards:
CRM adaptor layer - https://trello.com/c/hOldMbJT
CRM integration tests - https://trello.com/c/0F0zeAdO
Map CRM layer into JSON response - https://trello.com/c/pUy7Y17J

### CODE REVIEW

File: src/DqtApi/Controllers/TeachersController.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        public TeachersController(IDataverseAdaptor dataverseAdaptor) {
+            _dataverseAdaptor = dataverseAdaptor;
+        }
+
+        [HttpGet("{trn}")]
+        [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> GetTeacher(
+            [FromRoute] string trn,
+            [FromQuery(Name = "birthdate"), SwaggerParameter(Required = true)] string birthDate,  // TODO model binder for yyyy-mm-dd format
```
Comment (gunndabad) [inline src/DqtApi/Controllers/TeachersController.cs:29]: I think the built-in model binders will work with this date format
```suggestion
            [FromQuery(Name = "birthdate"), SwaggerParameter(Required = true)] DateTime? birthDate,
```

File: src/DqtApi/Models/Teacher.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Text.Json.Serialization;
+using Microsoft.Xrm.Sdk;
+
+namespace DqtApi.Models {
+
+    public class Teacher {
+        private Entity _entity;
+
+        // TODO fix attribute names - don't think these are currently right
```
Comment (gunndabad) [inline src/DqtApi/Models/Teacher.cs:10]: Is this comment stale?

File: src/DqtApi/Validators/GetTeacherValidator.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DqtApi.Models;
+using FluentValidation;
+
+namespace DqtApi.Validators
+{
+    public class GetTeacherValidator : AbstractValidator<GetTeacherRequest>
```
Comment (gunndabad) [inline src/DqtApi/Validators/GetTeacherValidator.cs:10]: Is this used?

File: tests/DqtApi.Tests/UnitTests/TeachersController.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+using DqtApi.DAL;
+using DqtApi.Models;
+using DqtApi.Responses;
+using Microsoft.AspNetCore.Mvc;
+using Moq;
+using Xunit;
+
+namespace DqtApi.Tests.UnitTests
+{
+    public class TeachersController
```
Comment (gunndabad) [inline tests/DqtApi.Tests/UnitTests/TeachersController.cs:13]: Could we merge the tests here with `GetTeacherTests`? There seems to be some cross-over.

File: src/DqtApi/Models/Teacher.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Text.Json.Serialization;
+using Microsoft.Xrm.Sdk;
+
+namespace DqtApi.Models {
+
+    public class Teacher {
```
Comment (gunndabad) [inline src/DqtApi/Models/Teacher.cs:7]: Is this class required?

File: src/DqtApi/Validators/GetTeacherValidator.cs (hunk 3 above)
Comment (JonathanWacks) [reply src/DqtApi/Validators/GetTeacherValidator.cs:10]: It's going to be - leaving it in for now

File: src/DqtApi/Controllers/TeachersController.cs (hunk 1 above)
Comment (JonathanWacks) [reply src/DqtApi/Controllers/TeachersController.cs]: Applied default binding and re-pushed

File: src/DqtApi/Models/Teacher.cs (hunk 2 above)
Comment (JonathanWacks) [reply src/DqtApi/Models/Teacher.cs]: Removed this class, as it's unnecessary

File: src/DqtApi/DAL/MoreThanOneMatchingTeacherException.cs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+using System;
+
+namespace DqtApi.DAL {
```
[the remark this answers is not in the document]
Comment (JonathanWacks) [reply src/DqtApi/DAL/MoreThanOneMatchingTeacherException.cs]: No, removed

File: src/DqtApi/Models/Teacher.cs (hunk 5 above)
Comment (JonathanWacks) [reply src/DqtApi/Models/Teacher.cs]: No, removed it

File: tests/DqtApi.Tests/UnitTests/TeachersController.cs (hunk 4 above)
Comment (JonathanWacks) [reply tests/DqtApi.Tests/UnitTests/TeachersController.cs:13]: Not sure what cross-over there is?  DataverseIntegration\GetTeacherTests tests the DataverseAdaptor - I'll rename it to DataverseAdaptorTests.  DqtApi.FunctionalTests\Endpoints\GetTeacherTests tests end-to-end functionality.

File: src/DqtApi/Controllers/TeachersController.cs (hunk 1 above)
Comment (JonathanWacks) [reply src/DqtApi/Controllers/TeachersController.cs]: This caused a few tests to fail - issue is that default binding allows other date formats, but they're American, so e.g. birthdate=02-07-1996 will be allowed, but will map to Feb 7th not July 2nd.  If we want to force the yyyy-MM-dd format, I think we need a custom model binder.  Reverting to string for now.

**Review by gunndabad — APPROVED** [review]

**Pull request merged**